### ISSUE TITLE
Questions regarding non-string use

### ISSUE THREAD

**Issue by lylemoffitt** [BurntSushi/fst#37]
I was thinking about writing an FST based [point-in-poly](https://en.wikipedia.org/wiki/Point_in_polygon) algorithm. This is partially for fun, partially to test the applicability (and potential performance) of FST solutions to other classic computing problems. I am aware that it's probably [not ideal](http://blog.burntsushi.net/transducers/#not-a-general-purpose-data-structure) for what I have in mind, but I thought it'd be an interesting test anyway. In thinking about how I would set about doing this I came up with some questions I thought best to get the answers to before continuing. I understand that this library probably wasn't meant to be used for what I have in mind, I'm just trying to gauge how difficult the problem is and how much work I'm setting for myself before I get too deep.

Problem Domain
---

My problem initially concerns how to store the polygon's information. This might be as 2D coordinates, vectors, or included triangles. Preprocessing the polygon will probably be necessary. I intend to write a query automaton (based on `fst::Levenshtein`) to answer point-in-poly queries. I understand that I may also need to overload (or partially rewrite) some of the `fst::raw::*` modules in order to achieve this functionality or to better suit the data structure to storing what I have in mind. 

Questions:
----
1) Assume that we're storing numeric tuples, representing a coordinate or vector: what is the relative tradeoff between different encoding methods? In the [primes example](https://gist.github.com/anonymous/61ae8b305e6abff3291c) you use big-endian byte order to preserve lexicographic ordering. I could do the same for integer or floating-point values alone, but what about when stored together as a struct? A JSON representation obviates this issue and the lexicographic issue, but probably makes matters worse elsewhere.

2) How important is lexicographic ordering? Would any other strict-weak-ordering work? How extensively is this assumption encoded in the library? 

3) How important is it that nodes only store 1 byte? Could this requirement be relaxed such that every node stores N-bytes instead? How extensively is this assumption encoded in the library? If it's not possible, would it be sensible for the automaton to implement a look-ahead or backtrack operation?

4) Assume the coordinates for a polygon are preprocessed and stored such that the FST contains coordinate sequences for [monotone](https://en.wikipedia.org/wiki/Monotone_polygon), [polygon chains](https://en.wikipedia.org/wiki/Polygonal_chain). This would mean that a point is [inside the polygon](https://en.wikipedia.org/wiki/Point_location#Monotone_subdivisions) if two chains can be found which bound (a linear comparison) the point. Could an automaton be constructed to resolve this query?

5) Assume the FST is storing JSON encoded coordinate pairs. How would one write an automaton that could be used to retrieve all the strings in the set which represent points within a given *linear* distance of a provided point?

**Comment by BurntSushi** [comment]
I don't have time to respond to everything, but this question stuck out to me.

> How important is lexicographic ordering? Would any other strict-weak-ordering work? How extensively is this assumption encoded in the library?

I think it would be worth your time to read this section of my blog post on the topic to understand how this library works internally: http://blog.burntsushi.net/transducers/#finite-state-machines-as-data-structures

(Lexicographic ordering is manifest in how the FST is encoded.)

**Comment by lylemoffitt** [comment]
@BurntSushi 
Thank you for citing a section of your blog article I might have not fully grokked. I had read it before (and several of the papers you cited), but on second reading I believe the reason you are referring to is, that under [construction](http://blog.burntsushi.net/transducers/#construction) you need to be able to form a Trie (prefix-tree) out of the data before you can perform other DFA minimizations. Lexicographic ordering is important here because it provides a total ordering on each of the constituent *bytes* (characters) that is consistent with the ordering of the the datum (string) as a whole. DFA minimization based on incremental "freezing" is built on this behavior, and not having that characteristic would fundamentally impede your ability to process, compress, and store the FSA in memory. 

So the answer (as I understand it) seems to be "yes" to some extent. Any strict-weak-ordering defined on the bytes that make up your data (strings) could be used. Ordering of the string as a whole you get "for free" just because of the nature of the former. Am I still missing something?

----

Of course, this still requires some acknowledgement of Q3. Up to now, I've been saying that the FSA nodes "store" the bytes of the datum (one each). This isn't accurate, strictly speaking. My understanding from Daciuk and Weiss [paper](http://www.cs.put.poznan.pl/dweiss/site/publications/download/fsacomp.pdf) on FSA compression is that this is not the case. It's not clear to me from reading your source code if this (fig. 5, p. 10) is the storage layout you are using or not, so please point out if any of this doesn't apply. If so, than it makes sense that your nodes be byte-oriented, because this reduces the permutation space for all the different byte value you could be storing and allows you to better take advantage of statistical methods of compression. Furthermore, it minimizes the overflow when the byte isn't one of the 31 most common.

That being said, it doesn't seem like a byte-orientation is strictly necessary. A larger storage size is possible; it would just reduce the gains from this statistical method of compression. It would also reduce the gains of compression in general as you are more likely to have frequently repeated values (or partial values) that are not deduplicated. You would also need a comparison (ordering) function that is strict/weak on the larger value domain of whatever you are storing. It should even be possible to store variable-width values in the nodes using encoding techniques like those for v-coded address or UTF8's code-points. The current representation for the FST nodes is already variable width, so this wouldn't really have any greater impact beyond taking a little longer to decode than a single byte would.

However, even if you could store larger values in the nodes, it's not clear to me that this would really gain you anything. By the fundamental nature of computers, all values we could ever possibly want to express can be represented as a sequence of bytes. The only thing that larger values really get us is a more intuitive mapping between values stores and the *meaning* those values represent. 

For example, you can create an FST set of [IEEE-754 8-byte floating-point](https://www.wikiwand.com/en/Double-precision_floating-point_format) pairs, but looking at the individual bytes won't really have any meaning. That is, the FSA that makes sense of the bytes, won't be able to do so on a byte-wise level. It will encounter bytes that contain both sign and partial exponent, partial exponent and partial mantissa, and partials of the remainder of a mantissa. It could still concatenate those values in memory and make sense of it when it reaches the end, but it could not perform a query that relied on an understand of the full value of the float until then. You can only build the automaton to perform a query that can be resolved as a sequence of comparisons on arbitrary bit-slices of your data.

-----

So far, it still seems pretty possible that the FST could theoretically hold any type of value/key. It just needs a slightly-more generic API. I'm still new to Rust, but I believe the following should work.

For a FST to hold keys of type `K` the following must be true:

1. The key must have a *aggregate* storage representation.  This is satisfied if the traits `From<R>` and `Into<R>` are implemented for `K`, where `R` is one of the following representations:
    - A fixed-size slice of storage:`[S;N]`, where `N` is some size and `S` is the storage type (probably a new-type wrapper over `u8`).
    - A variable-sized slice of storage: `[S]`, where the length is determined from the data.
    - A tuple / struct `(T_1,T_2, ... T_n)`, where each type is convertible to `R` as defined above. 

2. A node may contain any storage type `S` as long as the following traits are implemented for it:
    - `Ord` is required so that nodes may be compared for the purposes of FST construction.
    - `Copy` is required to denote that storage types may be stored / retrieved from Memory Map arbitrarily.
    - `ByteOrder` is required to denote that a clear binary representation must exist for the storage type. This may alternatively be satisfied with implementations of `From<[u8;N]>` and `Into<[u8;N]>` for `S` (the default would be `N=1`).


With this interface it would then be possible to store a tuple with heterogenous types and have the FST be type-aware in its comparisons during construction. Querying FSAs would also gain this benefit automatically. 

---

TL;DR - The typesystem can make generic, aggregate, data storage in an FST possible. Multiple solutions exist.

**Comment by BurntSushi** [comment]
OK, you've written a lot. Let me just try to take a crack at answering some of your questions.

> Assume that we're storing numeric tuples, representing a coordinate or vector: what is the relative tradeoff between different encoding methods? In the primes example you use big-endian byte order to preserve lexicographic ordering. I could do the same for integer or floating-point values alone, but what about when stored together as a struct? A JSON representation obviates this issue and the lexicographic issue, but probably makes matters worse elsewhere.

JSON seems unnecessary, but I agree it would work. If you have a tuple of numbers, then it seems like doing a big-endian encoding of both and storing them adjacent to each other as the key would just work. This should generalize to any arbitrary arity.

I think the trade off here is that, generally, longer keys lead to longer look up times since lookup will never to traverse more states.

> How important is lexicographic ordering? Would any other strict-weak-ordering work? How extensively is this assumption encoded in the library?

I tried to answer this before, but I think the short answer is that this library is very heavily coupled with a byte-based representation, and therefore, a lexicographic ordering. It's been a while since I've carefully looked at the representation of states in memory, but I wouldn't be surprised if the various compression tricks relied quite heavily on the small alphabet size.

> How important is it that nodes only store 1 byte? Could this requirement be relaxed such that every node stores N-bytes instead? How extensively is this assumption encoded in the library? If it's not possible, would it be sensible for the automaton to implement a look-ahead or backtrack operation?

The 1 byte restriction I think is pretty heavily encoded, both in construction and in lookup. I can't immediately think of a reason why this restriction is necessary, but it is simpler to code and I'd question whether doing it differently would lead to anything more than marginal benefits. (A long tail of nodes in the current construction will, on typical data, consume exactly one byte per node anyway.) For example, if you stuff `foo` into a single node and later try to add `fobar`, then you'd have to unpack the `foo` node into two nodes. (This seems *tricky* but not impossible.)

> Assume the coordinates for a polygon are preprocessed and stored such that the FST contains coordinate sequences for monotone, polygon chains. This would mean that a point is inside the polygon if two chains can be found which bound (a linear comparison) the point. Could an automaton be constructed to resolve this query?

I'm not sure. I'm having a little trouble understanding the question. I think if I were you, I would just try to write out the automaton and see where it leads. I might be able to help more effectively at that point. :-)

> Assume the FST is storing JSON encoded coordinate pairs. How would one write an automaton that could be used to retrieve all the strings in the set which represent points within a given linear distance of a provided point?

The JSON aspect of this makes it hard for me to think about. If I think about it as two adjacent big-endian integers, then my natural question is: what is the relationship between the lexicographic ordering of coordinates and their distance?

> Lexicographic ordering is important here because it provides a total ordering on each of the constituent bytes (characters) that is consistent with the ordering of the the datum (string) as a whole. DFA minimization based on incremental "freezing" is built on this behavior, and not having that characteristic would fundamentally impede your ability to process, compress, and store the FSA in memory.

Exactly.

> So the answer (as I understand it) seems to be "yes" to some extent. Any strict-weak-ordering defined on the bytes that make up your data (strings) could be used. Ordering of the string as a whole you get "for free" just because of the nature of the former. Am I still missing something?

Yes. It's worth pointing out that the ordering of the string is, itself, lexicographic. By itself, this isn't actually meaningful. But, it turns out, computer scientists are clever and have built text encodings such that a lexicographic order does something sensible. For example, for all Unicode codepoints `p` and `q`, WLOG, if `p < q` then `utf8(p) < utf8(q)`. This is precisely the property that makes FSTs on strings useful, because there is a clear correspondence between the lexicographic order of a string and something that is abstractly useful. I think your task is to find this correspondence for your geometry problems. :-) (Or, at least, this is the single biggest piece that confounds me given your problem statement.)

N.B. While lexicographic ordering of UTF-8 strings yields something useful, it turns out that Unicode defines its own collation order that isn't "just lexicographic order." So lexicographic order is a bit of a cheat here, and just so happens to work well in a lot of cases (especially if you only care about English).

> Of course, this still requires some acknowledgement of Q3. Up to now, I've been saying that the FSA nodes "store" the bytes of the datum (one each). This isn't accurate, strictly speaking. My understanding from Daciuk and Weiss paper on FSA compression is that this is not the case. It's not clear to me from reading your source code if this (fig. 5, p. 10) is the storage layout you are using or not, so please point out if any of this doesn't apply. If so, than it makes sense that your nodes be byte-oriented, because this reduces the permutation space for all the different byte value you could be storing and allows you to better take advantage of statistical methods of compression. Furthermore, it minimizes the overflow when the byte isn't one of the 31 most common.

The line between "storing nodes" and "storing transitions" has always been blurry to me, and I don't quite understand the trade offs between them. I think in this library, the code as written tends to operate on nodes.

> However, even if you could store larger values in the nodes, it's not clear to me that this would really gain you anything. By the fundamental nature of computers, all values we could ever possibly want to express can be represented as a sequence of bytes. The only thing that larger values really get us is a more intuitive mapping between values stores and the meaning those values represent.

This is kind of what I surmised when thinking about what the public API would be.

> So far, it still seems pretty possible that the FST could theoretically hold any type of value/key. It just needs a slightly-more generic API. I'm still new to Rust, but I believe the following should work.

"theoretically" being the operative word. I think you'll find a stronger coupling to bytes than you might expect. For example, I don't see anything about alphabets in your generic API, which probably immediately prevents some forms of compression.

My summary:

I think your geometry problem is really cool. I'd love to remain a soundboard for you. I think you have two paths before you:

1. Find a lexicographic mapping that is useful for the problem you're trying to solve. (Example: big-endian encoding for integers is useful. A little-endian encoding would also work---the FST itself wouldn't care---but interaction with the FST would be a lot less useful if your task is to show or query numbers in a particular range.)
2. Build out your generic version in a separate crate and see where that takes you. I would only go this route if you find this particular rabbit hole more appetizing than the geometry problems you're trying to solve. :-)
